File: src/Integration.Tests/python/test_optional.py
from typing import Any, Optional, Tuple

def test_int(n: Optional[int]) -> Optional[int]:
    return n

def test_str(s: Optional[str]) -> Optional[str]:
    return s

def test_any(obj: Optional[Any]) -> Optional[Any]:
    return obj

def test_optional_tuple(a: Optional[Tuple[Optional[int], Optional[str]]]) -> Optional[Tuple[Optional[int], Optional[str]]]:
    return a

File: src/Integration.Tests/TestOptional.cs
using CSnakes.Runtime.Python;

namespace Integration.Tests;
public class TestOptional(PythonEnvironmentFixture fixture) : IntegrationTestBase(fixture)
{
    [Theory]
    [InlineData(null)]
    [InlineData(42)]
    public void Int(int? input)
    {
        var mod = Env.TestOptional();
        Assert.Equal(input, mod.TestInt(input));
    }

    [Theory]
    [InlineData(null)]
    [InlineData("foobar")]
    public void Str(string? input)
    {
        var mod = Env.TestOptional();
        Assert.Equal(input, mod.TestStr(input));
    }

    [Theory]
    [InlineData(true)]
    [InlineData(false)]
    public void Object(bool @null)
    {
        var input = @null ? null : PyObject.Zero;
        var mod = Env.TestOptional();
        Assert.Equal(input, mod.TestAny(input));
    }

    [Fact]
    public void Tuple()
    {
        (int?, string) t = (10, "hello");
        var mod = Env.TestOptional();
        var result = mod.TestOptionalTuple(t);
        Assert.Equal(t, result);
    }
}
